tools/quantization/quantize.py

```diff
@@ -0,0 +1,899 @@
+            self.weight_qrange = 254  # [-(2^{b-1}-1), 2^{b-1}-1]: [-127, 127] for 8 bits.
+        else:
+            raise ValueError('unsupported quantization data type')
+
+        if self.input_qType == onnx_proto.TensorProto.UINT8:
+            self.input_qrange = 255 # 2^b - 1
+        else:
+            raise ValueError('unsupported quantization data type')
+
+        # QuantizeRange tensor name and zero tensor name for scale and zero point calculation.
+        # Used when QuantizationMode.is_static_mode() is False
+        self.fixed_qrange_name = "fixed_quantization_range"
+        self.fixed_zero_name = "fixed_zero"
+
+    def quantize_model(self):
+        # Create a new topologically sorted list for quantizing a model
+        new_list = []
+        for node in self.model.graph.node:
+            if node.op_type == 'Conv':
+                new_list += self._quantize_convolution(node, new_list)
+            elif node.op_type == 'MatMul':
+                new_list += self._quantize_matmul(node, new_list)
+            else:
+                new_list.append(node)
+
+        # extend is used to append to the list for a protobuf fields
+        # https://developers.google.com/protocol-buffers/docs/reference/python-generated?csw=1#fields
+        self.model.graph.ClearField('node')
+        self.model.graph.node.extend(new_list)
+
+        # update opset.
+        opset_info = next((opset for opset in self.model.opset_import if opset.domain == '' or opset.domain == onnx_domain), None)
+        if opset_info is not None:
+            self.model.opset_import.remove(opset_info)
+        self.model.opset_import.extend([onnx.helper.make_opsetid(onnx_domain, onnx_version)])
+
+        return self.model
+
+    def find_weight_data(self, initializer):
+        '''
+            :param initializer: TensorProto initializer object from a graph
+            :return: a list of initialized data in a given initializer object
+        '''
+        if initializer.data_type == onnx_proto.TensorProto.FLOAT:
+            weights = onnx.numpy_helper.to_array(initializer)
+        else:
+            raise ValueError('Model contains conv operator weights in {}. Only float type quantization is supported.'.format(
+                type_to_name[initializer.data_type]))
+        return weights
+
+    def _update_graph(self, weight):
+        '''
+            Given a weight object, update the graph by doing the following:
+             - remove old initializer, update new initializers for quantized weight, zero point, and scale
+             - remove old weight input, update with new inputs for quantized weight, zero point, and scale
+            This function does NOT update the nodes in the graph, just initializers and inputs
+        '''
+        packed_weight_name = weight.name + '_quantized'
+        scale_name = weight.name + '_scale'
+        zero_point_name = weight.name + '_zero_point'
+
+        # Remove existing weight initializer
```

Comment:
What happens when this input is also being used by another node? This condition should be checked.